proposed/simple-csharp-programs-motivation-walkthrough.md

```diff
@@ -0,0 +1,311 @@
+
+### Coding against a framework/sdk like ASP.NET Core
+
+You can also reference a framework/sdk (nomenclature up for debate) via a shebang:
+
+```csharp
+#!/usr/bin/share/dotnet-aspnet
+
+var app = WebApplication.CreateBuilder(args);
+app.MapGet("/", () => "Hello World");
+await app.RunAsync();
+```
+
+The `dotnet-aspnet` gesture indicates two things to the build/run mechanism:
+
+* ASP.NET Core gets loaded and referenced
+* Applicable implicit ASP.NET Core `using`s are brought into scope
+
+And you can run it:
+
+```bash
+$ dotnet run
+Listening on localhost://12345
+```
+
+The syntax for coding against a framework/sdk could also be more similar to the `#r "nuget:"` syntax shown earlier instead of being a shebang.
+
+## `dotnet` command support
+
+Only `dotnet run` and `dotnet build` are supported with Simple C# Programs. All other `dotnet` commands will error out, indicating that you need a project.
+
+## Grow-up story
+
+There is eventually a need for a grow-up story. What's interesting is that this can, in theory, go one of two ways:
+
+1. You "graduate" to using projects, like every other .NET developer
+2. You continue to create loose files, but you instead organize them in meaningful folders
+
+The first option is well-understood today, but could involve a special `dotnet` command that _adds_ a project rather than wipes out your existing directory, and converts special directives in your file:
+
+```bash
+$ dotnet add project
+    Added project 'MyFolder.csproj' to the current workspace
+```
+
+This lets you "grow up" into a project by:
+
+* Creating a project file and placing it in the current directory, with a name that matches
+* Potentially removing shebangs / special directives and converting things to an appropriate SDK attribute or `PackageReference`
+* Provides a nice message saying what it did
+
+After this, you need to run your program with `dotnet run` and friends, but by virtue of being a project, you now have the full `dotnet` toolset at your disposal.
+
+Shebangs wouldn't technically need to be removed. You should still be able to run a single file the same was as before. Having a project file in the current directory shouldn't change that.
+
+The second option has potentially wild implications across our entire build tools and IDE tooling stack. It would require us to fundamentally rethink how .NET applications can be built.
+
+For the purposes of this design, we will require that users create a project-based codebase if they wish to be grown up.
+
+## Design points / pivots
+
+The following is a non-exhuastive list of user experience considerations that go beyond the basic walkthrough above.
+
+### Multiple files
+
+We will need to decide if Simple C# Programs support more than one file. In other languages, like JavaScript and F#, you can have several loose files in your directory (or other directories) and include them into your program. For example:
```

Comment:
What about the case of multiple files where each file is a separate/independent program? For example, what if I have a directory with (at least) 2 files:

- `check-server-status.cs`
- `reboot-server.cs`

Both of these would be independent stand-alone programs that can be executed individually via `./check-server-status.cs` and `./reboot-server.cs`. Would dotnet be able to build/run them as such? Or is it assumed that if there are multiple files they are related and need to be built together and only one can be executed?

Reply:
I suspect that we'd need an explicit directive like `#load` in F# or `require` in JS. Otherwise each would be assumed be a separate program, and you'd need to specify which to run (e.g., `dotnet run reboot-server.cs`). I'll likely add this to the walkthrough itself, but I expect this to be a point of debate.